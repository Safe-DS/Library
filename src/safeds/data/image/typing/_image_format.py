from enum import Enum


class ImageFormat(Enum):
    """Images formats supported by us."""

    JPEG = "jpeg"
    PNG = "png"
