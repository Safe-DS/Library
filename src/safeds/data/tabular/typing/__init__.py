"""Types used to define the schema of a tabular dataset."""

from ._column_type import Anything, Boolean, ColumnType, Integer, RealNumber, String
from ._imputer_strategy import ImputerStrategy
from ._schema import Schema

__all__ = [
    "Anything",
    "Boolean",
    "ColumnType",
    "ImputerStrategy",
    "Integer",
    "RealNumber",
    "Schema",
    "String",
]
