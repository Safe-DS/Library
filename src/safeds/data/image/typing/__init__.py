"""Types used to distinguish different image formats."""

from ._image_format import ImageFormat

__all__ = [
    "ImageFormat",
]
