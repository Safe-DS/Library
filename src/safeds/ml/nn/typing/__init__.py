"""Types used to define neural networks and related attributes."""

from typing import TYPE_CHECKING

import apipkg

if TYPE_CHECKING:
    from ._model_image_size import ConstantImageSize, ModelImageSize, VariableImageSize

apipkg.initpkg(
    __name__,
    {
        "ConstantImageSize": "._model_image_size:ConstantImageSize",
        "ModelImageSize": "._model_image_size:ModelImageSize",
        "VariableImageSize": "._model_image_size:VariableImageSize",
    },
)

__all__ = [
    "ConstantImageSize",
    "ModelImageSize",
    "VariableImageSize",
]
