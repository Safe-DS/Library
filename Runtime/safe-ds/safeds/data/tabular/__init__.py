from ._column import Column, ColumnStatistics
from ._row import Row
from ._table import Table
