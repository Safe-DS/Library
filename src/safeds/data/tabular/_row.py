import typing
from typing import Any

import pandas as pd
from IPython.core.display_functions import DisplayHandle, display
from safeds.data.tabular.typing import ColumnType, TableSchema
from safeds.exceptions import UnknownColumnNameError


class Row:
    def __init__(self, data: typing.Iterable, schema: TableSchema):
        self._data: pd.Series = data if isinstance(data, pd.Series) else pd.Series(data)
        self.schema: TableSchema = schema
        self._data = self._data.reset_index(drop=True)

    def __getitem__(self, column_name: str) -> Any:
        return self.get_value(column_name)

    def __iter__(self) -> typing.Iterator[Any]:
        return iter(self.get_column_names())

    def __len__(self) -> int:
        return len(self._data)

    def get_value(self, column_name: str) -> Any:
        """
        Return the value of a specified column.

        Parameters
        ----------
        column_name : str
            The column name.

        Returns
        -------
        value :
            The value of the column.
        """
        if not self.schema.has_column(column_name):
            raise UnknownColumnNameError([column_name])
        return self._data[self.schema._get_column_index_by_name(column_name)]

    def count(self) -> int:
        """
        Return the number of columns in this row.

        Returns
        -------
        count : int
            The number of columns.
        """
        return len(self._data)

    def has_column(self, column_name: str) -> bool:
        """
        Return whether the row contains a given column.
        Alias for self.schema.hasColumn(column_name: str) -> bool.

        Parameters
        ----------
        column_name : str
            The name of the column.

        Returns
        -------
        contains : bool
            True, if row contains the column.
        """
        return self.schema.has_column(column_name)

    def get_column_names(self) -> list[str]:
        """
        Return a list of all column names saved in this schema.
        Alias for self.schema.get_column_names() -> list[str].

        Returns
        -------
        column_names : list[str]
            The column names.
        """
        return self.schema.get_column_names()

    def get_type_of_column(self, column_name: str) -> ColumnType:
        """
        Return the type of a specified column.
        Alias for self.schema.get_type_of_column(column_name: str) -> ColumnType.

        Parameters
        ----------
        column_name : str
            The name of the column.

        Returns
        -------
        type : ColumnType
            The type of the column.

        Raises
        ------
        ColumnNameError
            If the specified target column name does not exist.
        """
        return self.schema.get_type_of_column(column_name)

    def __eq__(self, other: typing.Any) -> bool:
        if not isinstance(other, Row):
            return NotImplemented
        if self is other:
            return True
        return self._data.equals(other._data) and self.schema == other.schema

    def __hash__(self) -> int:
        return hash(self._data)

    def __str__(self) -> str:
        tmp = self._data.to_frame().T
        tmp.columns = self.get_column_names()
        return tmp.__str__()

    def __repr__(self) -> str:
        tmp = self._data.to_frame().T
        tmp.columns = self.get_column_names()
        return tmp.__repr__()

    def _ipython_display_(self) -> DisplayHandle:
        """
        Return a display object for the column to be used in Jupyter Notebooks.

        Returns
        -------
        output : DisplayHandle
            Output object.
        """
        tmp = self._data.to_frame().T
        tmp.columns = self.get_column_names()

        with pd.option_context(
            "display.max_rows", tmp.shape[0], "display.max_columns", tmp.shape[1]
        ):
            return display(tmp)
