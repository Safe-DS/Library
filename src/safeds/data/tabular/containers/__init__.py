from ._column import Column
from ._row import Row
from ._table import Table
from ._tagged_table import TaggedTable
