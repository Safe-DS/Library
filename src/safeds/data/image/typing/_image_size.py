from __future__ import annotations

import sys
from typing import TYPE_CHECKING

from safeds._utils import _structural_hash
from safeds.exceptions import ClosedBound, OutOfBoundsError

if TYPE_CHECKING:
    from safeds.data.image.containers import Image


class ImageSize:
    """
    A container for image size data.

    Parameters
    ----------
    width:
        the width of the image
    height:
        the height of the image
    channel:
        the channel of the image

    Raises
    ------
    OutOfBoundsError:
        if width or height are below 1
    ValueError
        if an invalid channel is given
    """

    def __init__(self, width: int, height: int, channel: int, *, _ignore_invalid_channel: bool = False) -> None:
        if width < 1 or height < 1:
            raise OutOfBoundsError(min(width, height), lower_bound=ClosedBound(1))
        elif not _ignore_invalid_channel and channel not in (1, 3, 4):
            raise ValueError(f"Channel {channel} is not a valid channel option. Use either 1, 3 or 4")
        elif channel < 1:
            raise OutOfBoundsError(channel, name="channel", lower_bound=ClosedBound(1))
        self._width = width
        self._height = height
        self._channel = channel

    @staticmethod
    def from_image(image: Image) -> ImageSize:
        """
        Create a `ImageSize` of a given image.

        Parameters
        ----------
        image:
            the given image for the `ImageSize`

        Returns
        -------
        image_size:
            the calculated `ImageSize`
        """
        return ImageSize(image.width, image.height, image.channel)

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, ImageSize):
            return NotImplemented
        return (self is other) or (
            self._width == other._width and self._height == other._height and self._channel == other._channel
        )

    def __hash__(self) -> int:
        return _structural_hash(self._width, self._height, self._channel)

    def __sizeof__(self) -> int:
        return sys.getsizeof(self._width) + sys.getsizeof(self._height) + sys.getsizeof(self._channel)

    def __str__(self) -> str:
        return f"{self._width}x{self._height}x{self._channel} (WxHxC)"

    @property
    def width(self) -> int:
        """
        Get the width of this `ImageSize` in pixels.

        Returns
        -------
        width:
            The width of this `ImageSize`.
        """
        return self._width

    @property
    def height(self) -> int:
        """
        Get the height of this `ImageSize` in pixels.

        Returns
        -------
        height:
            The height of this `ImageSize`.
        """
        return self._height

    @property
    def channel(self) -> int:
        """
        Get the channel of this `ImageSize` in pixels.

        Returns
        -------
        channel:
            The channel of this `ImageSize`.
        """
        return self._channel
