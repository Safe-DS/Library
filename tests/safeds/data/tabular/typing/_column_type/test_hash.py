from collections.abc import Callable

import pytest
from safeds.data.tabular.typing import ColumnType
from syrupy import SnapshotAssertion


@pytest.mark.parametrize(
    "type_factory",
    [
        lambda: ColumnType.float32(),
        lambda: ColumnType.float64(),
        lambda: ColumnType.int8(),
        lambda: ColumnType.int16(),
        lambda: ColumnType.int32(),
        lambda: ColumnType.int64(),
        lambda: ColumnType.uint8(),
        lambda: ColumnType.uint16(),
        lambda: ColumnType.uint32(),
        lambda: ColumnType.uint64(),
        lambda: ColumnType.date(),
        lambda: ColumnType.datetime(),
        lambda: ColumnType.duration(),
        lambda: ColumnType.time(),
        lambda: ColumnType.string(),
        lambda: ColumnType.binary(),
        lambda: ColumnType.boolean(),
        lambda: ColumnType.null(),
    ],
    ids=[
        "float32",
        "float64",
        "int8",
        "int16",
        "int32",
        "int64",
        "uint8",
        "uint16",
        "uint32",
        "uint64",
        "date",
        "datetime",
        "duration",
        "time",
        "string",
        "binary",
        "boolean",
        "null",
    ],
)
class TestSameHash:
    def test_should_return_same_hash_for_equal_column_types(self, type_factory: Callable[[], ColumnType]) -> None:
        type_1 = type_factory()
        type_2 = type_factory()
        assert hash(type_1) == hash(type_2)

    def test_should_return_same_hash_in_different_processes(
        self,
        type_factory: Callable[[], ColumnType],
        snapshot: SnapshotAssertion,
    ) -> None:
        type_ = type_factory()
        assert hash(type_) == snapshot


@pytest.mark.parametrize(
    ("type_1", "type_2"),
    [
        (ColumnType.float32(), ColumnType.float64()),
        (ColumnType.float32(), ColumnType.int32()),
        (ColumnType.int32(), ColumnType.uint32()),
        (ColumnType.int32(), ColumnType.string()),
    ],
    ids=[
        "different bit count",
        "float vs. int",
        "signed vs. unsigned",
        "numeric vs. non-numeric",
    ],
)
def test_should_return_different_hash_for_unequal_column_types(
    type_1: ColumnType,
    type_2: ColumnType,
) -> None:
    assert hash(type_1) != hash(type_2)
