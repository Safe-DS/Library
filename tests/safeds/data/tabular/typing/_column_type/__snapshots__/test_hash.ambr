# serializer version: 1
# name: TestSameHash.test_should_return_same_hash_in_different_processes[binary]
  9114564521908245695
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[boolean]
  253312121440042057
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[date]
  1420129173814739165
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[datetime]
  5037917593634747378
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[duration]
  61885111053208984
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[float32]
  2208906555873308475
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[float64]
  7992127945058937926
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[int16]
  231909494783576033
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[int32]
  1374055615029084484
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[int64]
  1212246054477521308
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[int8]
  2269551742465714004
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[null]
  1357169886076964811
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[string]
  1529874693733179990
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[time]
  5418105405405492344
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[uint16]
  937488211706751048
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[uint32]
  859241000506496144
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[uint64]
  1728989395548489096
# ---
# name: TestSameHash.test_should_return_same_hash_in_different_processes[uint8]
  48638156943199247
# ---
